src/Components/Components/src/Routing/Router.cs

```diff
@@ -190,28 +192,31 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
-        private async Task RunOnNavigateAsync(string path)
+        private async Task<bool> RunOnNavigateAsync(string path, Task prevOnNavigate)
         {
             // If this router instance does not provide an OnNavigateAsync parameter
             // then we render the component associated with the route as per usual.
             if (!OnNavigateAsync.HasDelegate)
             {
-                return;
+                return true;
             }
 
             // If we've already invoked a task and stored its CTS, then
-            // cancel the existing task.
-            _onNavigateCts?.Dispose();
+            // cancel that existing task.
+            _onNavigateCts?.Cancel();
+            // Then make sure that the task has been completed cancelled or
+            // completed before continuing with the execution of this current task.
+            await prevOnNavigate;
 
             // Create a new cancellation token source for this instance
             _onNavigateCts = new CancellationTokenSource();
             var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
 
             // Create a cancellation task based on the cancellation token
             // associated with the current running task.
-            var cancellationTaskSource = new TaskCompletionSource();
+            var cancellationTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Because the previous OnNavigate is always awaited, am I right that there can only be one call in progress?

How much do you care about performance? You might be able to change the TaskCompletionSource here to use a task backed by https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.manualresetvaluetasksourcecore-1?view=dotnet-plat-ext-3.1. That would basically allow for a resettable completion source.

Reply:
> Because the previous OnNavigate is always awaited, am I right that there can only be one call in progress?

Correct.

> You might be able to change the TaskCompletionSource here to use a task backed by https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.manualresetvaluetasksourcecore-1?view=dotnet-plat-ext-3.1. That would basically allow for a resettable completion source.

Interesting. I wasn't aware of this API. Might be worthwhile to explore. Strikes me as a bit more readable to use here.

Reply:
I played around with this. I like not having to do the allocation for a new TCS at every invocation but the API for MRVTSC is a little meh in some places (e.g. no way to set task creation options in a constructor since its a struct).

For now, I'll avoid prematurely optimizing this and see if users run into perf issues when using this in practice. Thanks for the recommendation, tho!

Reply:
I wouldn't worry about an allocation in this case, seems like a premature optimization.